### PR TITLE
[App/Feature] HTTP Queues

### PR DESCRIPTION
**Pull request by hhsecond**
## What does this PR do?

Introducing HTTPQueues to Lightning. This brings in the new HTTPQueue class and the HTTPClient classes. It also makes changes to the cloud dispatcher to pass the queue type. The queue type is defaults to "redis" but for those who wants to use HTTPQueue can set the environmental variable for that. Once the HTTPQueue is stabilized, we'll swap the default to HTTPQueue

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/Lightning-AI/lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [x] Did you make sure to **update the documentation** with your changes? (if necessary)
- [x] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [x] Did you **update the CHANGELOG**? (not for typos, docs, test updates, or minor internal changes/refactors)

## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/Lightning-AI/lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [ ] Is this pull request ready for review? (if not, please submit in draft mode)
- [ ] Check that all items from **Before submitting** are resolved
- [ ] Make sure the title is self-explanatory and the description concisely explains the PR
- [ ] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃

cc @borda

### CODE REVIEW

**Comment by hhsecond** [comment]
Dependency PRs on the Lightning Cloud hasn't been merged yet. Please don't merge this PR until then

File: src/lightning_app/core/queues.py [hunk 1]
```diff
@@ -291,3 +301,98 @@ def is_running(self) -> bool:
...
+        if timeout is None:
+            while True:
+                try:
+                    return self._get()
+                except queue.Empty:
+                    time.sleep(HTTP_QUEUE_REFRESH_INTERVAL)
+
+        # make one request and return the result
+        if timeout == 0:
+            return self._get()
```
Comment (tchaton) [inline src/lightning_app/core/queues.py:339]: Should this one be wrapped into a try catch, what happened if the service doesn't answer ?

File: src/lightning_app/core/queues.py [hunk 2]
```diff
@@ -291,3 +301,98 @@ def is_running(self) -> bool:
...
+
+    def _get(self):
+        resp = self.client.post(f"v1/{self.app_id}/{self.name}", query_params={"action": "pop"})
+        if resp.status_code == 204:
+            raise queue.Empty
+        return pickle.loads(resp.content)
+
+    def put(self, item: Any) -> None:
+        if not self.app_id:
+            raise ValueError(f"App ID couldn't be extracted from the queue name: {self._original_name}")
```
Comment (tchaton) [inline src/lightning_app/core/queues.py:358]:
```suggestion
            raise ValueError(f"The Lightning App ID couldn't be extracted from the queue name: {self._original_name}")
```

File: src/lightning_app/core/queues.py [hunk 3]
```diff
@@ -291,3 +301,98 @@ def is_running(self) -> bool:
...
+
+        value = pickle.dumps(item)
+        queue_len = self.length()
+        if queue_len >= WARNING_QUEUE_SIZE:
+            warnings.warn(
+                f"The Queue {self.name} length is larger than the recommended length of {WARNING_QUEUE_SIZE}. "
+                f"Found {queue_len}. This might cause your application to crash, please investigate this."
+            )
+        resp = self.client.post(f"v1/{self.app_id}/{self.name}", data=value, query_params={"action": "push"})
```
Comment (tchaton) [inline src/lightning_app/core/queues.py:367]: Should this include projects?

**Review by tchaton — APPROVED** [review]
LGTM !

File: src/lightning_app/core/queues.py (hunk 3 above)
Comment (hhsecond) [reply src/lightning_app/core/queues.py:367]: We don't really need to include the project ID technically, as far as app ID is unique across the system and this isn't really following REST, so no real need to arrange the resources hierarchically.

File: src/lightning_app/core/queues.py (hunk 1 above)
Comment (hhsecond) [reply src/lightning_app/core/queues.py:339]: We are implicitly retrying for 2 minutes for most of the failure modes (except the service completely goes down or if we get a connection refused error). I have a TODO somewhere in the code to address this in the future (before making HTTPQueue the default Queue)

What's the behaviour you'd expect if the service is down? We'll retry forever, in a defined frequency?

**Review by manskx — APPROVED** [review]
🚀

File: src/lightning_app/core/queues.py (hunk 1 above)
Comment (hhsecond) [reply src/lightning_app/core/queues.py:339]: I am going to keep this for a future PR

**Pull request merged**